ocaml/xapi/xapi_psr.ml

```diff
@@ -0,0 +1,143 @@
+
+  type host
+
+  val save_checkpoint : string -> unit
+
+  val retrieve_checkpoint : unit -> string option
+
+  val backup : pool_secrets -> unit
+
+  val retrieve : unit -> pool_secrets
+
+  val tell_accept_new_pool_secret : pool_secrets -> host -> unit
+
+  val tell_send_new_pool_secret : pool_secrets -> host -> unit
+
+  val tell_cleanup_old_pool_secret : host -> unit
+
+  val cleanup_master : unit -> unit
+end
+
+module Make =
+functor
+  (Impl : Impl)
+  ->
+  struct
+    open Impl
+
+    type checkpoint =
+      | No_checkpoint (* ==> previous rotation was successful*)
+      | Accept_new_pool_secret
+      | Send_new_pool_secret
+      | Cleanup_slaves
+      | Cleanup_master
+    [@@deriving rpcty]
+
+    exception R of host r
+
+    exception Cannot_parse_checkpoint of string
+
+    let string_of_checkpoint x =
+      Rpcmarshal.marshal checkpoint.Rpc.Types.ty x |> Jsonrpc.to_string
+      |> (* remove leading and trailing '"' *)
+      fun s -> String.sub s 1 (String.length s - 2)
+
+    let checkpoint_of_string s =
+      let rpc = Rpc.rpc_of_string s in
+      match Rpcmarshal.unmarshal checkpoint.Rpc.Types.ty rpc with
+      | Ok x ->
+          x
+      | Error _ ->
+          raise (Cannot_parse_checkpoint s)
+
+    let rec go pool_secrets hosts = function
+      | No_checkpoint ->
+          (* if we fail to backup the pool secrets, it doesn't really matter,
+             you can simply restart the rotation *)
+          backup pool_secrets ;
+          (go [@tailcall]) pool_secrets hosts Accept_new_pool_secret
+      | Accept_new_pool_secret ->
+          save_checkpoint (string_of_checkpoint Accept_new_pool_secret) ;
+          let () =
+            hosts
+            |> List.iter (fun host ->
+                   try tell_accept_new_pool_secret pool_secrets host
+                   with e ->
+                     D.error
+                       "failed while telling hosts to accept new pool secrets. \
+                        error= %s"
+                       (Printexc.to_string e) ;
+                     raise (R (Failed_during_accept_new_pool_secret host)))
+          in
+          (go [@tailcall]) pool_secrets hosts Send_new_pool_secret
+      | Send_new_pool_secret ->
+          save_checkpoint (string_of_checkpoint Send_new_pool_secret) ;
+          let () =
+            hosts
+            |> List.iter (fun host ->
+                   try tell_send_new_pool_secret pool_secrets host
+                   with e ->
+                     D.error
+                       "failed while telling hosts to send new pool secret. \
+                        error= %s"
+                       (Printexc.to_string e) ;
+                     raise (R (Failed_during_send_new_pool_secret host)))
+          in
+          (go [@tailcall]) pool_secrets hosts Cleanup_slaves
+      | Cleanup_slaves ->
+          save_checkpoint (string_of_checkpoint Cleanup_slaves) ;
+          let[@warning "-8"] (master :: slaves) = hosts in
```

Comment:
Perhaps have `master` and `slaves` as arguments to `go`, so you don't need this?